src/Functions/bitSlice.cpp

```diff
@@ -0,0 +1,476 @@
+            if (length < 0)
+            {
+                length_byte = std::max(remain_byte + (length / word_size), static_cast<ssize_t>(0));
+                over_bit = word_size + (length % word_size);
+                if (length_byte == 1 && over_bit <= offset_bit) // begin and end are in same byte AND there are no gaps
+                    length_byte = 0;
+            }
+            bool right_truncate = static_cast<ssize_t>(length_byte) > remain_byte;
+            size_t abandon_last_bit = (over_bit && !right_truncate) ? word_size - over_bit : 0;
+            auto slice = src.getSliceFromLeft(offset_byte, length_byte);
+            if (slice.size)
+                writeSliceWithLeftShift(slice, sink, shift_bit, abandon_last_bit);
+
+            sink.next();
+            src.next();
+        }
+    }
+
+
+    template <class Source>
+    void bitSliceFromRightConstantOffsetBounded(Source && src, StringSink && sink, size_t offset, ssize_t length) const
+    {
+        size_t offset_byte = offset / word_size;
+        size_t offset_bit = (word_size - (offset % word_size)) % word_size; // offset_bit always represent left offset bit
+        if (offset_bit)
+            offset_byte++;
+        size_t length_byte = 0;
+        size_t over_bit = 0;
+        if (length > 0)
+        {
+            length_byte = (length + offset_bit) / word_size;
+            over_bit = (length + offset_bit) % word_size;
+            if (over_bit && (length_byte || over_bit > offset_bit)) // begin and end are not in same byte OR there are gaps
+                length_byte++;
+        }
+
+        while (!src.isEnd())
+        {
+            size_t size = src.getElementSize();
+            if (length < 0)
+            {
+                length_byte = std::max(static_cast<ssize_t>(offset_byte) + (length / word_size), static_cast<ssize_t>(0));
+                over_bit = word_size + (length % word_size);
+                if (length_byte == 1 && over_bit <= offset_bit) // begin and end are in same byte AND there are no gaps
+                    length_byte = 0;
+            }
+            bool left_truncate = offset_byte > size;
+            bool right_truncate = length_byte > offset_byte;
+            size_t shift_bit = left_truncate ? 0 : offset_bit;
+            size_t abandon_last_bit = (over_bit && !right_truncate) ? word_size - over_bit : 0;
+            auto slice = src.getSliceFromRight(offset_byte, length_byte);
+            if (slice.size)
+                writeSliceWithLeftShift(slice, sink, shift_bit, abandon_last_bit);
+
+            sink.next();
+            src.next();
+        }
+    }
+
+    template <class Source>
+    void bitSliceDynamicOffsetBounded(Source && src, StringSink && sink, const IColumn & offset_column, const IColumn & length_column) const
+    {
+        const bool is_offset_null = offset_column.onlyNull();
+        const auto * offset_nullable = typeid_cast<const ColumnNullable *>(&offset_column);
```

Comment:
Do we want to handle nulls? If yes in what way? According to code below we replace nulls with some defaults. But `useDefaultImplementationForNulls` is not set, so nulls rows with nulls in args will result to null.
Also same question for useDefaultImplementationForConstants

Reply:
`bitSlice(s, offset, lenght)`, I think null lenght and offset values should not make result to null. The behavior here is consistent with `substring` and `arrayslice`.

Reply:
You are right, There is no need to handle nulls. I will remove these code.